pulp_gem/specs.py

```diff
@@ -149,52 +149,166 @@ async def read_info(relative_path, versions):
+        "specification_version",
+        "name",
+        "version",
+        "date",
+        "summary",
+        "required_ruby_version",
+        "required_rubygems_version",
+        "original_platform",
+        "dependencies",
+        "rubyforge_project",  # This got removed...
+        "email",
+        "authors",
+        "description",
+        "homepage",
+        "has_rdoc",
+        "new_platform",
+        "licenses",
+        "metadata",
     ]
-    # write uncompressed version
-    with open(relative_path, "wb") as fd:
-        rubymarshal.writer.write(fd, specs)
+
+    @classmethod
+    def yaml_constructor(cls, loader, node):
+        result = cls()
+        yield result
+
+        value = loader.construct_mapping(node, deep=True)
+        platform = value.pop("platform")
+        value.setdefault("original_platform", platform)
+        value.setdefault("new_platform", platform)
+        value.setdefault("has_rdoc", True)
+        value.setdefault("rubyforge_project", "")
+        value["date"] = RubyTime.from_datetime(value["date"])
+        result._private_data = {key: value.get(key) for key in cls.FIELDS}
+
+    def _load(self, data):
+        arguments = rubymarshal.reader.loads(data)
+        self._private_data = {name: value for name, value in zip(self.FIELDS, arguments)}
+
+    def _dump(self):
+        count = len(self._private_data)
+        arguments = [
+            self._private_data[name] if name is not None else "" for name in self.FIELDS[:count]
+        ]
+        return rubymarshal.writer.writes(arguments)
+
+
+class GemRequirement(rubymarshal.classes.UsrMarshal):
+    ruby_class_name = "Gem::Requirement"
+
+    @classmethod
+    def yaml_constructor(cls, loader, node):
+        result = cls()
+        yield result
+        values = loader.construct_mapping(node, deep=True)
+        result.marshal_load([values["requirements"]])
+
+    @property
+    def requirements(self):
+        return self._private_data[0]
+
+    def __repr__(self):
+        return f"{self.ruby_class_name}('{self.requirements}')"
+
+    def __str__(self):
+        return f"{self.ruby_class_name}('{self.requirements}')"
+
+    def __eq__(self, other):
+        return isinstance(other, self.__class__) and self._private_data == self._private_data
+
+    def to_s(self):
+        return "&".join([f"{req[0]} {req[1].version}" for req in self.requirements])
+
+
+class GemDependency(rubymarshal.classes.RubyObject):
+    ruby_class_name = "Gem::Dependency"
+
+    @classmethod
+    def yaml_constructor(cls, loader, node):
+        result = cls()
+        yield result
+        values = loader.construct_mapping(node, deep=True)
+        result.attributes = {f"@{key}": value for key, value in values.items()}
+        result.attributes["@type"] = rubymarshal.classes.Symbol(result.attributes["@type"][1:])
+
+
+class RubyTime(rubymarshal.classes.UserDef):
+    ruby_class_name = "Time"
+
+    @classmethod
+    def from_datetime(cls, value):
```

Comment:
Is their a library we can get this conversion from? This code is scary looking.

Reply:
I don't think so. This Looks like really specific to rubymarshal. The best bet is probably to push these classes towards the library we have: python-rubymarshal.

pulp_ansible/tests/functional/api/collection/test_marks.py:from